### ISSUE TITLE
gef-remote address miscalculation with cross target gdb

### ISSUE THREAD

**Issue by eternaleclipse** [hugsy/gef#175]
#### Host gdb

gdb with gef: Ubuntu Linux x86_64, gdb and gdbserver cloned and compiled from here:
https://android.googlesource.com/toolchain/gdb.git

```
./configure --with-python=/usr/bin/python3.5 --target=arm-linux-androideabi
make
sudo make install
```

Python version:
```
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609] on linux
```

#### Target gdbserver

Nexus 5, Android 6.0 with cyanogenmod
```Linux localhost 3.4.0-g2aa165e #1 SMP PREEMPT Thu Aug 20 06:07:34 UTC 2015 armv7l```

gdbserver compiled with:
```
CC=arm-linux-androideabi-gcc ./configure --target=arm-linux-androideabi --host=arm-linux-androideabi
make
```

### The problem

When using gdb's native command ```target remote <ip>:<port>``` everything works fine.
I am able to ```disassemble main``` and it shows the correct address and instructions.

When using ```gef-remote <ip>:<port>``` and ```disassemble main```, for somewhat reason it probably miscalculates (or fails to retrieve correctly) the base address as 0x0 and when it tries to access ```main``` it adds 0x0 + 0x47c and gets to inaccessible memory.

#### Steps to reproduce

  1. Compile cross-target gdb and from android repo as explained earlier
  2. Cross-compile gdbserver as explained earlier
  3. Upload and run gdbserver on target and connect with gef-remote
  4. Try to run ```disassemble main```

#### Observed Results

```
gef➤  disas main 
Dump of assembler code for function main:
   0x0000047c <+0>:     Cannot access memory at address 0x47c
```

#### Expected results

```
gef➤  disas main 
Dump of assembler code for function main:
   0x2a00047c <+0>:     push    {r11, lr}
   0x2a000480 <+4>:     add     r11, sp, #4
   0x2a000484 <+8>:     sub     sp, sp, #24
   0x2a000488 <+12>:    sub     r3, r11, #24
   0x2a00048c <+16>:    mov     r0, r3
   0x2a000490 <+20>:    bl      0x2a0003b0 <gets@plt>
   0x2a000494 <+24>:    mov     r3, #0
   0x2a000498 <+28>:    mov     r0, r3
   0x2a00049c <+32>:    sub     sp, r11, #4
   0x2a0004a0 <+36>:    pop     {r11, pc}
End of assembler dump.
gef➤  
```

#### Traces
Demonstration of how it works nicely with gdb's native ```target remote```:
```
user@ubuntu ~ % arm-linux-androideabi-gdb
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "--host=x86_64-pc-linux-gnu --target=arm-linux-androideabi".
---Type <return> to continue, or q <return> to quit---
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
GEF for linux ready, type `gef' to start, `gef config' to configure
51 commands loaded for GDB 7.11 using Python engine 3.5
[*] 5 commands could not be loaded, run `gef missing` to know why.
gef➤  target remote 10.100.102.2:5000                                                                                                                                                                       
Remote debugging using 10.100.102.2:5000
Reading /data/local/tmp/bof from remote target...
warning: File transfers from remote targets can be slow. Use "set sysroot" to access files locally instead.
Reading /data/local/tmp/bof from remote target...
Reading symbols from target:/data/local/tmp/bof...(no debugging symbols found)...done.
Reading /system/bin/linker from remote target...
Reading /system/bin/linker from remote target...
Reading symbols from target:/system/bin/linker...(no debugging symbols found)...done.
0xb6fdf654 in __dl__start () from target:/system/bin/linker
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────[ registers ]────
$r0   : 0x00000000
$r1   : 0xbefffc28  →  0x00666f62 ("bof"?)
$r2   : 0x00000000
$r3   : 0x00000000
$r4   : 0x00000000
$r5   : 0x00000000
$r6   : 0x00000000
$r7   : 0x00000000
$r8   : 0x00000000
$r9   : 0x00000000
$r10  : 0x00000000
$r11  : 0x00000000
$r12  : 0x00000000
$sp   : 0xbefffb20  →  0x00000001
$lr   : 0x00000000
$pc   : 0xb6fdf654  →  <__dl__start+0> mov r0,  sp
$cpsr : [thumb fast interrupt overflow carry zero negative]
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────[ stack ]────
0xbefffb20│+0x00: 0x00000001     ← $sp
0xbefffb24│+0x04: 0xbefffc28  →  0x00666f62 ("bof"?)
0xbefffb28│+0x08: 0x00000000
0xbefffb2c│+0x0c: 0xbefffc2c  →  "_=./gdbserver"
0xbefffb30│+0x10: 0xbefffc3a  →  "ANDROID_DATA=/data"
0xbefffb34│+0x14: 0xbefffc4d  →  "MKSH=/system/bin/sh"
0xbefffb38│+0x18: 0xbefffc61  →  "TMPDIR=/data/local/tmp"
0xbefffb3c│+0x1c: 0xbefffc78  →  "LOGNAME=root"
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────[ code:armv5te ]────
   0xb6fdf63c <__dl__GLOBAL__sub_I_stubs.cpp+36> ldr??.w r0,  [r0,  r1]
   0xb6fdf640 <__dl__GLOBAL__sub_I_stubs.cpp+40> vst4.8 {d16-d19},  [r6],  r1
   0xb6fdf644 <__dl__GLOBAL__sub_I_pthread_atfork.cpp+0> ldr    r3,  [pc,  #8]  ; (0xb6fdf650 <__dl__GLOBAL__sub_I_pthread_atfork.cpp+12>)
   0xb6fdf646 <__dl__GLOBAL__sub_I_pthread_atfork.cpp+2> movs   r2,  #0
   0xb6fdf648 <__dl__GLOBAL__sub_I_pthread_atfork.cpp+4> add    r3,  pc
   0xb6fdf64a <__dl__GLOBAL__sub_I_pthread_atfork.cpp+6> str    r2,  [r3,  #0]
 → 0xb6fdf654 <__dl__start+0>  mov    r0,  sp
   0xb6fdf658 <__dl__start+4>  blx    0xb6fe40d8 <__dl___linker_init>
   0xb6fdf65c <__dl__start+8>  mov    pc,  r0
   0xb6fdf660 <__dl__ZL24debuggerd_signal_handleriP7siginfoPv+0> stmdb  sp!,  {r4,  r5,  r6,  r7,  r8,  r9,  r10,  r11,  lr}
   0xb6fdf664 <__dl__ZL24debuggerd_signal_handleriP7siginfoPv+4> sub    sp,  #196       ; 0xc4
   0xb6fdf666 <__dl__ZL24debuggerd_signal_handleriP7siginfoPv+6> add.w  r8,  sp,  #80   ; 0x50
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────[ threads ]────
[#0] Id 1, Name: "bof", stopped, reason: SIGTRAP
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────[ trace ]────
[#0] 0xb6fdf654 → Name: __dl__start()
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
gef➤  disas main 
Dump of assembler code for function main:
   0x2a00047c <+0>:     push    {r11, lr}
   0x2a000480 <+4>:     add     r11, sp, #4
   0x2a000484 <+8>:     sub     sp, sp, #24
   0x2a000488 <+12>:    sub     r3, r11, #24
   0x2a00048c <+16>:    mov     r0, r3
   0x2a000490 <+20>:    bl      0x2a0003b0 <gets@plt>
   0x2a000494 <+24>:    mov     r3, #0
   0x2a000498 <+28>:    mov     r0, r3
   0x2a00049c <+32>:    sub     sp, r11, #4
   0x2a0004a0 <+36>:    pop     {r11, pc}
End of assembler dump.
gef➤  
```

Demonstration of the problem with gef-remote:
```
user@ubuntu ~ % arm-linux-androideabi-gdb
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "--host=x86_64-pc-linux-gnu --target=arm-linux-androideabi".
---Type <return> to continue, or q <return> to quit---
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
GEF for linux ready, type `gef' to start, `gef config' to configure
51 commands loaded for GDB 7.11 using Python engine 3.5
[*] 5 commands could not be loaded, run `gef missing` to know why.
gef➤  gef-remote 10.100.102.2:5000                                                                                                                                                                          
Reading /data/local/tmp/bof from remote target...
warning: File transfers from remote targets can be slow. Use "set sysroot" to access files locally instead.
Reading /data/local/tmp/bof from remote target...
Reading symbols from target:/data/local/tmp/bof...(no debugging symbols found)...done.
Reading /system/bin/linker from remote target...
Reading /system/bin/linker from remote target...
0xb6fdf654 in __dl__start () from target:/system/bin/linker
[+] Connected to '10.100.102.2:5000'
[+] Targeting PID=3852
[+] Downloading remote information
[+] Remote information loaded, remember to clean '/tmp/gef/3852' when your session is over
gef➤  disassemble main 
Dump of assembler code for function main:
   0x0000047c <+0>:     Cannot access memory at address 0x47c
gef➤  
```

**Comment by hugsy** [comment]
Tested with `gdb-multiarch` Debian packet, and cannot reproduce (addresses solved correctly). Did you try with it too ?

**Comment by hugsy** [comment]
Btw this should incidentally be fixed by d898c55. I will close this ticket when #210 is merged.

**Comment by hugsy** [comment]
Closing this ticket which was fixed with #210
